File: VB/CodeUtils/ExampleFinder.vb
Imports System
Imports System.Collections.Generic
Imports System.IO
Imports System.Text.RegularExpressions

Namespace RichEditDocumentServerAPIExample.CodeUtils

#Region "ExampleFinder"
    Public MustInherit Class ExampleFinder

        Public MustOverride ReadOnly Property RegexRegionPattern As String

        Public MustOverride ReadOnly Property RegionStartPattern As String

        Public MustOverride ReadOnly Property RegionHelperStartPattern As String

        Public Function Process(ByVal fileWithExample As FileInfo) As List(Of CodeExample)
            If fileWithExample Is Nothing Then Return New List(Of CodeExample)()
            Dim groupName As String = Path.GetFileNameWithoutExtension(fileWithExample.Name)
            Dim code As String
            Using stream As FileStream = File.Open(fileWithExample.FullName, FileMode.Open, FileAccess.Read)
                Dim sr As StreamReader = New StreamReader(stream)
                code = sr.ReadToEnd()
            End Using

            Dim foundExamples As List(Of CodeExample) = ParseSouceFileAndFindRegionsWithExamples(groupName, code)
            Return foundExamples
        End Function

        Public Function ParseSouceFileAndFindRegionsWithExamples(ByVal groupName As String, ByVal sourceCode As String) As List(Of CodeExample)
            Dim result As List(Of CodeExample) = New List(Of CodeExample)()
            Dim matches = Regex.Matches(sourceCode, RegexRegionPattern, RegexOptions.Singleline)
            For Each match In matches
                Dim lines As String() = match.ToString().Split(New String() {Microsoft.VisualBasic.Constants.vbCrLf, Microsoft.VisualBasic.Constants.vbLf}, StringSplitOptions.None)
                If lines.Length <= 2 Then Continue For
                lines = DeleteLeadingWhiteSpacesFromSourceCode(lines)
                Dim regionName As String = String.Empty
                Dim regionIsValid As Boolean = ValidateRegionName(lines, regionName)
                If Not regionIsValid Then Continue For
                Dim exampleCode As String = String.Join(Microsoft.VisualBasic.Constants.vbCrLf, lines, 1, lines.Length - 2)
                result.Add(CreateRichEditExample(groupName, regionName, exampleCode))
            Next

            Return result
        End Function

        Protected Function CreateRichEditExample(ByVal exampleGroup As String, ByVal regionName As String, ByVal exampleCode As String) As CodeExample
            Dim result As CodeExample = New CodeExample()
            SetExampleCode(exampleCode, result)
            result.RegionName = regionName
            result.HumanReadableGroupName = ConvertStringToHumanReadableForm(exampleGroup)
            Return result
        End Function

        Protected MustOverride Sub SetExampleCode(ByVal exampleCode As String, ByVal newExample As CodeExample)

        Protected Overridable Function DeleteLeadingWhiteSpacesFromSourceCode(ByVal lines As String()) As String()
            Return DeleteLeadingWhiteSpaces(lines, "        ")
        End Function

        Protected Overridable Function ValidateRegionName(ByVal lines As String(), ByRef regionName As String) As Boolean
            Dim keepHashMark As Integer = 0 ' "#example" if value is -1 or "example" if value will be 0
            Dim region As String = lines(0)
            Dim regionIndex As Integer = region.IndexOf(RegionHelperStartPattern)
            If regionIndex = 0 Then
                regionName = ConvertStringToHumanReadableForm(region.Substring(regionIndex + RegionHelperStartPattern.Length + keepHashMark))
            End If

            If regionIndex < 0 Then
                regionIndex = region.IndexOf(RegionStartPattern)
                If regionIndex < 0 Then
                    regionName = String.Empty
                    Return False
                End If

                regionName = ConvertStringToHumanReadableForm(region.Substring(regionIndex + RegionStartPattern.Length + keepHashMark))
            End If

            Return True
        End Function
    End Class

#End Region
#Region "ExampleFinderVB"
    Public Class ExampleFinderVB
        Inherits ExampleFinder

        Public Overrides ReadOnly Property RegexRegionPattern As String
            Get
                Return "#Region.*?#End Region"
            End Get
        End Property

        Public Overrides ReadOnly Property RegionStartPattern As String
            Get
                Return "#Region ""#"
            End Get
        End Property

        Public Overrides ReadOnly Property RegionHelperStartPattern As String
            Get
                Return "#Region ""#@"
            End Get
        End Property

        Protected Overrides Function DeleteLeadingWhiteSpacesFromSourceCode(ByVal lines As String()) As String()
            Dim result As String() = MyBase.DeleteLeadingWhiteSpacesFromSourceCode(lines)
            Return CodeExampleUtils.DeleteLeadingWhiteSpaces(result, Microsoft.VisualBasic.Constants.vbTab & Microsoft.VisualBasic.Constants.vbTab)
        End Function

        Protected Overrides Function ValidateRegionName(ByVal lines As String(), ByRef regionName As String) As Boolean
            Dim result As Boolean = MyBase.ValidateRegionName(lines, regionName)
            If Not result Then Return result
            regionName = regionName.TrimEnd(""""c)
            Return True
        End Function

        Protected Overrides Sub SetExampleCode(ByVal code As String, ByVal newExample As CodeExample)
            newExample.CodeVB = code
        End Sub
    End Class

#End Region
#Region "ExampleFinderCSharp"
    Public Class ExampleFinderCSharp
        Inherits ExampleFinder

        Public Overrides ReadOnly Property RegexRegionPattern As String
            Get
                Return "#region.*?#endregion"
            End Get
        End Property

        Public Overrides ReadOnly Property RegionStartPattern As String
            Get
                Return "#region #"
            End Get
        End Property

        Public Overrides ReadOnly Property RegionHelperStartPattern As String
            Get
                Return "#region #@"
            End Get
        End Property

        Protected Overrides Sub SetExampleCode(ByVal code As String, ByVal newExample As CodeExample)
            newExample.CodeCS = code
        End Sub
    End Class
#End Region
End Namespace

File: CS/CodeUtils/ExampleFinder.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text.RegularExpressions;

namespace RichEditDocumentServerAPIExample.CodeUtils
{
    #region ExampleFinder
    public abstract class ExampleFinder
    {
        public abstract string RegexRegionPattern { get; }
        public abstract string RegionStartPattern { get; }
        public abstract string RegionHelperStartPattern { get; }

        public List<CodeExample> Process(FileInfo fileWithExample)
        {
            if (fileWithExample == null)
                return new List<CodeExample>();

            string groupName = Path.GetFileNameWithoutExtension(fileWithExample.Name);
            string code;
            using (FileStream stream = File.Open(fileWithExample.FullName, FileMode.Open, FileAccess.Read))
            {
                StreamReader sr = new StreamReader(stream);
                code = sr.ReadToEnd();
            }
            List<CodeExample> foundExamples = ParseSouceFileAndFindRegionsWithExamples(groupName, code);
            return foundExamples;
        }
        public List<CodeExample> ParseSouceFileAndFindRegionsWithExamples(string groupName, string sourceCode)
        {
            List<CodeExample> result = new List<CodeExample>();

            var matches = Regex.Matches(sourceCode, RegexRegionPattern, RegexOptions.Singleline);

            foreach (var match in matches)
            {
                string[] lines = match.ToString().Split(new string[] { "\r\n", "\n" }, StringSplitOptions.None);

                if (lines.Length <= 2)
                    continue;
                lines = DeleteLeadingWhiteSpacesFromSourceCode(lines);

                string regionName = String.Empty;
                bool regionIsValid = ValidateRegionName(lines, ref regionName);
                if (!regionIsValid)
                    continue;

                string exampleCode = string.Join("\r\n", lines, 1, lines.Length - 2);
                result.Add(CreateRichEditExample(groupName, regionName, exampleCode));
            }
            return result;
        }

        protected CodeExample CreateRichEditExample(string exampleGroup, string regionName, string exampleCode)
        {
            CodeExample result = new CodeExample();
            SetExampleCode(exampleCode, result);
            result.RegionName = regionName;
            result.HumanReadableGroupName = CodeExampleUtils.ConvertStringToHumanReadableForm(exampleGroup);
            return result;
        }
        protected abstract void SetExampleCode(string exampleCode, CodeExample newExample);

        protected virtual string[] DeleteLeadingWhiteSpacesFromSourceCode(string[] lines)
        {
            return CodeExampleUtils.DeleteLeadingWhiteSpaces(lines, "        ");
        }
        protected virtual bool ValidateRegionName(string[] lines, ref string regionName)
        {
            int keepHashMark = 0; // "#example" if value is -1 or "example" if value will be 0

            string region = lines[0];
            int regionIndex = region.IndexOf(RegionHelperStartPattern);

            if (regionIndex == 0)
            {
                regionName = CodeExampleUtils.ConvertStringToHumanReadableForm(region.Substring(regionIndex + RegionHelperStartPattern.Length + keepHashMark));
            }

            if (regionIndex < 0)
            {
                regionIndex = region.IndexOf(RegionStartPattern);
                if (regionIndex < 0)
                {
                    regionName = String.Empty;
                    return false;
                }
                regionName = CodeExampleUtils.ConvertStringToHumanReadableForm(region.Substring(regionIndex + RegionStartPattern.Length + keepHashMark));
            }
            return true;
        }
    }
    #endregion
    #region ExampleFinderVB
    public class ExampleFinderVB : ExampleFinder
    {
        public override string RegexRegionPattern { get { return "#Region.*?#End Region"; } }
        public override string RegionStartPattern { get { return "#Region \"#"; } }
        public override string RegionHelperStartPattern { get { return "#Region \"#@"; } }

        protected override string[] DeleteLeadingWhiteSpacesFromSourceCode(string[] lines)
        {
            string[] result = base.DeleteLeadingWhiteSpacesFromSourceCode(lines);
            return CodeExampleUtils.DeleteLeadingWhiteSpaces(result, "\t\t");
        }
        protected override bool ValidateRegionName(string[] lines, ref string regionName)
        {
            bool result = base.ValidateRegionName(lines, ref regionName);
            if (!result)
                return result;
            regionName = regionName.TrimEnd('\"');
            return true;
        }
        protected override void SetExampleCode(string code, CodeExample newExample)
        {
            newExample.CodeVB = code;
        }
    }
    #endregion
    #region ExampleFinderCSharp
    public class ExampleFinderCSharp : ExampleFinder
    {
        public override string RegexRegionPattern { get { return "#region.*?#endregion"; } }
        public override string RegionStartPattern { get { return "#region #"; } }
        public override string RegionHelperStartPattern { get { return "#region #@"; } }

        protected override void SetExampleCode(string code, CodeExample newExample)
        {
            newExample.CodeCS = code;
        }
    }
    #endregion
}
